torch/masked/maskedtensor/binary.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import torch
+
+from .core import _masks_match, _tensors_match, is_masked_tensor
+
+__all__ = []  # type: ignore[var-annotated]
+
+BINARY_NAMES = [
+    "add",
+    #    "addcdiv",
+    #    "addcmul",
+    "atan2",
+    "arctan2",
+    "bitwise_and",
+    "bitwise_or",
+    "bitwise_xor",
+    "bitwise_left_shift",
+    "bitwise_right_shift",
+    "div",
+    "divide",
+    "floor_divide",
+    "fmod",
+    "logaddexp",
+    "logaddexp2",
+    #    "logical_and",
+    #    "logical_or",
+    #    "logical_xor",
+    "mul",
+    "multiply",
+    "nextafter",
+    "remainder",
+    "sub",
+    "subtract",
+    "true_divide",
+    "eq",
+    "ne",
+    "le",
+    "ge",
+    "greater",
+    "greater_equal",
+    # "equal",
+    "gt",
+    # "isclose",
+    "less_equal",
+    "lt",
+    "less",
+    "maximum",
+    "minimum",
+    "fmax",
+    "fmin",
+    "not_equal",
+]
+
+INPLACE_BINARY_NAMES = [
+    n + "_"
+    for n in (
+        list(
+            set(BINARY_NAMES)
+            - {
+                "logaddexp",
+                "logaddexp2",
+                "equal",
+                "fmin",
+                "minimum",
+                "maximum",
+                "fmax",
+            }
+        )
+    )
+]
+
+
+def _get_at_least_one_mask(a, b):
+    if not is_masked_tensor(a) and not is_masked_tensor(b):
+        raise TypeError("At least one of `a` and `b` must be a MaskedTensor")
+    if not _masks_match(a, b):
+        raise ValueError("a and b must have matching masks")
+    if is_masked_tensor(a):
+        return a.get_mask()
+    return b.get_mask()
+
+
+def _binary_helper(fn, args, kwargs, inplace):
+    from .passthrough import _map_mt_args_kwargs, _wrap_result
+
+    if len(kwargs) != 0:
+        raise ValueError("len(kwargs) must equal 0")
+    if len(args) > 2:
+        for a in args[1:]:
+            if torch.is_tensor(a):
+                raise TypeError("MaskedTensor binary ops do not support Tensor arguments")
+
+    if not _masks_match(*args[:2]):
+        raise ValueError(
+            "Input masks must match. If you need support for this, please open an issue on Github."
+        )
+
+    if (
+        is_masked_tensor(args[0])
+        and is_masked_tensor(args[1])
+        and args[0].layout() != args[1].layout()
```

Comment:
yeah I think that's what the code below is relying on

Reply:
Sorry deleted my comment but you responded before I revised it! The code is meant to allow for lhs: MaskedTensor, rhs: dense tensor/float. The `is_masked_tensor` check at line 127 should read "if it has the same format"

Reply:
I think 127 moved since you commented :) also though, I'd suggest writing out the exact set of allowed layout combinations in a comment, and merging/deduplicating the logic here and in 105-109 into a single block. That'll make it much easier to understand/check the invariants that are being enforced here.